### DISCUSSION TITLE
Create a new repository based on PlatformPlatofrm but not a fork

### DISCUSSION THREAD

**Discussion by thiagolunardi** [platformplatform/PlatformPlatform#652] [General]
### Detailed description?

When starting a new project based on PlatformPlatform, I desire a simple and implicit way to keep my project foundation updated with the latest changes from PlatformPlatform without having it as a repository fork. That's because some restrictions apply to a forked repository, such as not allowing it to be private.

It will be convenient to have this tool provided by the `developer-cli`, something like:
```
pp create-based-repository --name MyPrivateRepository
```

This command should create, mirror, and clone a new repository on GitHub based on PlatformPlatform.

**Describe alternatives you've considered**

For my end, I did the following steps to achieve a good starting point:
1. Clone PlatformPlatform with `--bare` option
1. Create a new clean private repository on GitHub
1. From the cloned PlatformPlatform, do a `git push --mirror` to the private repository
1. Clone the new private repository and verify for the remote addresses with `git remove -v`, you should see something like this:
```
origin  https://github.com/<owner>/<private_repository>.git (fetch)
origin  https://github.com/<owner>/<private_repository>.git (push)
upstream  https://github.com/platformplatform/PlatformPlatform.git (fetch)
upstream  https://github.com/platformplatform/PlatformPlatform.git (push)
```

Now, we can update the private repository with changes from PlatformPlatform by doing:
```
 git fetch upstream
 git merge upstream
```

### Feature Type

DevOps, Reference example

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by tjementum** [comment]
This is actually a solved problem in PlatformPlatform, so converting this to a discussion ;)

We recently created a new CI command called `pull-platformplatform-changes` ([see this pull-request #642](https://github.com/platformplatform/PlatformPlatform/pull/642)). This command will detect any changes in the PlatformPlatform repository that are not implemented in a downstream project based on PlatformPlatform.

The CLI command will create a new branch called `platformplatform-updates` and then cherry-pick all pull request commits from PlatformPlatform into this branch, including the full pull request description. Then it will run all the unit tests, static code analysis, etc. to ensure nothing breaks before cherry-picking the next pull request from PlatformPlatform.

Also, there is a convention in PlatformPlatform pull requests descriptions that will show instructions to maintainers of downstream projects to make changes in their projects to align with the PlatformPlatform conventions, architecture, etc. Or, in the case of a truly breaking change in PlatformPlatform, it will stop the cherry pick with clear instructions on how to move forward.

Since PlatformPlatform is a boilerplate template, there will always be situations where there are merge conflicts. So, this is not a bulletproof way to pull changes, but it's much more structured than if maintainers of downstream projects manually have to use Git to pull changes.

FYI: Currently, there is a problem when running the `pull-platformplatform-changes` command on Windows because `npm` is a script and not an executable. This will be fixed very soon.

**Comment by thiagolunardi** [comment]
The CLI command can work for existing repositories. The case I mentioned would be to make the onboarding and adoption easier by giving one command to have a boosted head start on a brand-new repository.

I'm also wondering about creating a CLI command to add a new PlatformPlatform self-contained system. It would create a wireframe for the Core, API, web app, worker, tests and connect it to AppHost. Something like:

```
Description:
  Create a wireframe for a new self-contained system based on PlatformPlatofrm

Usage:
  pp new-self-container-system [options]

Options:
  --name       Name of the new self-contained system. E.g. file-management
  --with         Resource dependency list. Default: database, storage
  -?, -h, --help          Show help and usage information
```

New discussion?
